src/coprocessor/dag/batch_executor/table_scan_executor.rs

```diff
@@ -0,0 +1,906 @@
+    }
+
+    #[inline]
+    fn build_scanner<S: Store>(
+        &self,
+        store: &S,
+        desc: bool,
+        range: KeyRange,
+    ) -> Result<Scanner<S>> {
+        Ok(Scanner::new(
+            store,
+            crate::coprocessor::dag::ScanOn::Table,
+            desc,
+            self.key_only,
+            range,
+        )?)
+    }
+
+    /// Constructs empty columns, with PK in decoded format and the rest in raw format.
+    fn build_column_vec(&self, expect_rows: usize) -> LazyBatchColumnVec {
+        let columns_len = self.schema.len();
+        let mut columns = Vec::with_capacity(columns_len);
+
+        if let Some(handle_index) = self.handle_index {
+            // PK is specified in schema. PK column should be decoded and the rest is in raw format
+            // like this:
+            // non-pk non-pk non-pk pk non-pk non-pk non-pk
+            //                      ^handle_index = 3
+            //                                             ^columns_len = 7
+
+            // Columns before `handle_index` (if any) should be raw.
+            for _ in 0..handle_index {
+                columns.push(LazyBatchColumn::raw_with_capacity(expect_rows));
+            }
+            // For PK handle, we construct a decoded `VectorValue` because it is directly
+            // stored as i64, without a datum flag, at the end of key.
+            columns.push(LazyBatchColumn::decoded_with_capacity_and_tp(
+                expect_rows,
+                EvalType::Int,
+            ));
+            // Columns after `handle_index` (if any) should also be raw.
+            for _ in handle_index + 1..columns_len {
+                columns.push(LazyBatchColumn::raw_with_capacity(expect_rows));
+            }
+        } else {
+            // PK is unspecified in schema. All column should be in raw format.
+            for _ in 0..columns_len {
+                columns.push(LazyBatchColumn::raw_with_capacity(expect_rows));
+            }
+        }
+
+        LazyBatchColumnVec::from(columns)
+    }
+
+    fn process_kv_pair(
+        &mut self,
+        key: &[u8],
+        value: &[u8],
+        columns: &mut LazyBatchColumnVec,
+    ) -> Result<()> {
+        use crate::coprocessor::codec::{datum, table};
+        use crate::util::codec::number;
+
+        let columns_len = self.schema.len();
+        let mut decoded_columns = 0;
+
+        if let Some(handle_index) = self.handle_index {
+            let handle_id = table::decode_handle(key)?;
+            // TODO: We should avoid calling `push_int` repeatly. Instead we should specialize
+            // a `&mut Vec` first. However it is hard to program due to lifetime restriction.
+            columns[handle_index]
+                .mut_decoded()
+                .push_int(Some(handle_id));
+            decoded_columns += 1;
+            self.is_column_filled[handle_index] = true;
+        }
+
+        if value.is_empty() || (value.len() == 1 && value[0] == datum::NIL_FLAG) {
```

Comment:
Could we use `table::cut_row` or define an independent function to parse value to `col_id`=>`value`?

Reply:
We cannot use `table::cut_row` because it doesn't produce a data structure that we need.

We can define an independent function for this feature. But I don't think it is *very* necessary because I guess nowhere else will use it.

Reply:
I suggest to define an indepent function because:

> The first rule of functions is that they should be small. The second rule of functions is that they should be smaller than that.
>  In the eighties we used say that a function should be no bigger than a screen-full. Today Lines should not be 150 characters on a line and 100 lines or more on a screeen. Functions should hardly ever be 20 lines long.
> 
> --- clean code

Reply:
Why not use `!` to avoid the empty block?

Reply:
This makes it explicit that we are intended to do nothing in this case, not we missed the case.

Reply:
I think using `!` can represent the same meaning even if it is not explicit as the current implementation. And I am not sure if this is a good code style. But I am ok with it, just curious about it.

Reply:
@AndreMouche But looks like even if we extract it into a function, that function is still very long? How can we further resolve it?

Reply:
Ok